feat(linter): add `oxc-security/api-keys`

DonIsaac opened the pull request:
## What This PR does
Adds a new `oxc-security/api-key` rule that scans for hard-coded API keys.

It is broken up into "secret rules", where each one is responsible for finding a different kind of key. It is architecturally identical to how lint rules themselves. This PR also includes the first of these rules, for AWS access key IDs.

Logic and rules are based on [keyhunter](https://github.com/Donisaac/keyhunter). I've licensed that repo under GNU GPLv3, but it's my code and I can do what I want with it :smiling_imp: (read: I'm fine with it being MIT for oxc).


This PR is a complete feature in its own right, but does not represent the end of this work. See https://github.com/oxc-project/backlog/issues/116 to track overall progress.

DonIsaac commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5906?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5906** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5906?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#5905** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5905?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_linter/src/rules/security/api_keys/mod.rs:
```diff
@@ -0,0 +1,178 @@
+    /// This is a performance optimization to avoid checking each rule for every string.
+    min_len: NonZeroU32,
+    /// Minimum entropy over all enabled secret rules.
+    /// This is a performance optimization to avoid checking each rule for every string.
```
```suggestion
```

camc314 replied:
or maybe ignore me actually

camc314 commented on crates/oxc_linter/src/rules/security/api_keys/mod.rs:
```diff
@@ -0,0 +1,178 @@
+            return;
+        }
+
+        for rule in dbg!(&self.rules) {
```
dbg! ?

camc314 reviewed (approved):
nice!

DonIsaac committed:
feat(linter): add `oxc-security/api-keys` (#5906)

## What This PR does
Adds a new `oxc-security/api-key` rule that scans for hard-coded API keys.

It is broken up into "secret rules", where each one is responsible for finding a different kind of key. It is architecturally identical to how lint rules themselves. This PR also includes the first of these rules, for AWS access key IDs.

Logic and rules are based on [keyhunter](https://github.com/Donisaac/keyhunter). I've licensed that repo under GNU GPLv3, but it's my code and I can do what I want with it :smiling_imp: (read: I'm fine with it being MIT for oxc).

This PR is a complete feature in its own right, but does not represent the end of this work. See https://github.com/oxc-project/backlog/issues/116 to track overall progress.
